File: atcoder/abc/abc049/d/main.py
class UnionFind:
	def __init__(self, N):
		self.root = list(range(N + 1))
		self.size = [1] * (N + 1)
		self.N = N
		self.group_size = N
 
	def find_root(self, x):
		root = self.root
		while root[x] != x:
			root[x] = root[root[x]]
			x = root[x]
		return x
 
	def merge(self, x, y):
		root, size = self.root, self.size
		x = self.find_root(x)
		y = self.find_root(y)
		if x == y:
			return
		self.group_size -= 1
		if size[x] < size[y]:
			x, y = y, x
		root[y] = root[x]
		size[x] += size[y]
	
	def is_same(self, x, y):
		return self.find_root(x)==self.find_root(y)

	def group_size(self):
		return self.group_size

	def groups(self):
		G = [[] for _ in range(self.N)]
		for i in range(self.N):
			G[self.find_root(i)].append(i)
		return [i for i in G if i]
from collections import defaultdict
N, K, L = map(int, input().split())
train = UnionFind(N)
road = UnionFind(N)
for _ in range(K):
    p, q = map(int, input().split())
    p -= 1
    q -= 1
    road.merge(p, q)
for _ in range(L):
    p, q = map(int, input().split())
    p -= 1
    q -= 1
    train.merge(p, q)

count = defaultdict(int)
ans = [0]*N
for i in range(N):
    count[(road.find_root(i), train.find_root(i))] += 1

for i in range(N):
    ans[i] = count[(road.find_root(i), train.find_root(i))]
print(*ans)

File: atcoder/abc/abc049/d/main.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long
struct dsu{
  int N;
  vector<int> root, size;
  dsu(int _N=0): root(_N), size(_N){
  for(int i=0;i<_N;i++){
    root[i]=i;
  }
  }

  int merge(int x, int y){
    x = find(x);
    y = find(y);
    if (x==y) return -1;
    if (size[y]>size[x]) swap(x, y);
    root[y] = root[x];
    size[y] += size[x];
    return root[y];
  }

  bool same(int x, int y){
    return find(x)==find(y);
  }

  int find(int x){
    if (root[x]==x) return x;
    root[x] = find(root[x]);
    return root[x];
    
  }

  vector<vector<int>> groups(){
    vector<vector<int>> keep(N, vector<int>());
    vector<vector<int>> ret;
    for(int i=0;i<N;i++){
      keep[find(i)].push_back(i);
    }
    for(int i=0;i<N;i++){
      if(keep[i].size()!=0){
        ret.push_back(keep[i]);
      }
    }
    return ret;
  }
};

int main(){
  ll N, K, L;
  cin >> N >> K >> L;
  dsu road(N);
  dsu train(N);
  int u, v;
  for(int i=0;i<K;i++){
    cin >> u >> v;
    u--;
    v--;
    road.merge(u, v);
  }
  for(int i=0;i<L;i++){
    cin >> u >> v;
    u--;
    v--;
    train.merge(u, v);
  }

  vector<ll> ans(N, 0);
  unordered_map<ll, ll> cnt;
  for(int i=0;i<N;i++){
    ll a = road.find(i);
    ll b = train.find(i);
    cnt[a*N+b]++;
  }
  for(int i=0;i<N;i++){
    ll a = road.find(i);
    ll b = train.find(i);
    cout << cnt[a*N+b] << ' ';
  }

}